src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/AbstractReferenceHandler.cs

```diff
@@ -68,32 +71,73 @@ internal async Task<List<ProjectSystemReferenceInfo>> GetReferencesAsync(Configu
         {
+            var propertyNames = await metadata.GetPropertyNamesAsync();
             string? value = await metadata.GetEvaluatedPropertyValueAsync(ProjectReference.TreatAsUsedProperty);
 
             return value != null && PropertySerializer.SimpleTypes.ToValue<bool>(value);
         }
 
-        internal async Task<bool> UpdateReferenceAsync(ConfiguredProject selectedConfiguredProject, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+        private async Task<IProjectItem> GetProjectItems(ConfiguredProject selectedConfiguredProject,
+            string itemSpecification)
         {
-            bool wasUpdated = false;
-
-            cancellationToken.ThrowIfCancellationRequested();
-
             var projectItems = await GetUnresolvedReferencesAsync(selectedConfiguredProject);
 
             var item = projectItems
-                .FirstOrDefault(c => c.EvaluatedInclude == referenceUpdate.ReferenceInfo.ItemSpecification);
+                .FirstOrDefault(c => c.EvaluatedInclude == itemSpecification);
+            return item;
+        }
+
+        internal IProjectSystemUpdateReferenceOperation CreateSetAttributeCommand(ConfiguredProject selectedConfiguredProject,
+            ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            return new SetAttributeCommand(this, selectedConfiguredProject, referenceUpdate.ReferenceInfo.ItemSpecification);
+        }
+
+        internal IProjectSystemUpdateReferenceOperation CreateUnSetAttributeCommand(ConfiguredProject selectedConfiguredProject,
+            ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            return new UnSetAttributeCommand(this, selectedConfiguredProject, referenceUpdate.ReferenceInfo.ItemSpecification);
+        }
+
+        internal IProjectSystemUpdateReferenceOperation? CreateRemoveReferenceCommand(ConfiguredProject selectedConfiguredProject,
+            ProjectSystemReferenceUpdate referenceUpdate)
+        {
+            return new RemoveReferenceCommand(this, selectedConfiguredProject, referenceUpdate);
+        }
+
+        public async Task<Dictionary<string, string>> GetAttributesAsync(ConfiguredProject selectedConfiguredProject, string itemSpecification)
+        {
+            Dictionary<string, string> propertyValues = new ();
 
-            if (item != null)
+            IProjectItem items = await GetProjectItems(selectedConfiguredProject, itemSpecification);
+
+            var propertyNames = await items.Metadata.GetPropertyNamesAsync();
+
+            foreach (var property in propertyNames)
             {
-                string newValue = PropertySerializer.SimpleTypes.ToString(referenceUpdate.Action == ProjectSystemUpdateAction.SetTreatAsUsed);
+                var value = await items.Metadata.GetEvaluatedPropertyValueAsync(property);
+                propertyValues.Add(string.Copy(property), string.Copy(value));
```

Comment:
I can't see a reason to copy the strings here.

```suggestion
                propertyValues.Add(property, value);
```

In .NET strings are immutable, except for in uncommon corner cases that don't appear to apply here. If they do apply here, they should be called out in a comment.